sameboy.pc.in

```diff
@@ -0,0 +1,10 @@
+prefix=@PREFIX@
+exec_prefix=${prefix}
+includedir=${prefix}/include
+libdir=${exec_prefix}/lib
```

Comment:
`${exec_prefix}`, `${includedir}`, and `${libdir}` are not always inside of the `${prefix}` and `${libdir}` should not be hardcoded to `lib`, in slackware they use lib64 on x86_64 for instance.

Reply:
I've changed it like this:
```diff
modified   Makefile
@@ -46,6 +46,9 @@ ifneq ($(shell which xdg-open 2> $(NULL))$(FREEDESKTOP),)
 # Running on an FreeDesktop environment, configure for (optional) installation
 DESTDIR ?= 
 PREFIX ?= /usr/local
+EXEC_PREFIX ?= $(PREFIX)/bin
+INCLUDEDIR ?= $(PREFIX)/include
+LIBDIR ?= $(PREFIX)/lib
 DATA_DIR ?= $(PREFIX)/share/sameboy/
 FREEDESKTOP ?= true
 endif
@@ -822,7 +825,11 @@ $(LIBDIR)/libsameboy.so: $(LIBDIR)/libsameboy.o
 $(PKGCONF_FILE): sameboy.pc.in
 	-@$(MKDIR) -p $(dir $@)
 	-@rm -f $@
-	sed -e 's,@PREFIX@,$(PREFIX),' -e 's/@VERSION@/$(VERSION)/' $< > $@
+	sed -e 's,@PREFIX@,$(PREFIX),' \
+	 -e 's/@VERSION@/$(VERSION)/' \
+	 -e 's,@EXEC_PREFIX@,$(EXEC_PREFIX),' \
+	 -e 's,@INCLUDE_DIR@,$(INCLUDE_DIR),' \
+	 -e 's,@LIBDIR@,$(LIBDIR),' $< > $@
 
 pkgconf: $(PKGCONF_FILE)
 
modified   sameboy.pc.in
@@ -1,7 +1,7 @@
 prefix=@PREFIX@
-exec_prefix=${prefix}
-includedir=${prefix}/include
-libdir=${exec_prefix}/lib
+exec_prefix=@EXEC_PREFIX@
+includedir=@INCLUDE_DIR@
+libdir=@LIBDIR@
 
 Name: sameboy
 Description: The SameBoy library
```

Reply:
That is not quite what I mean.

In the pkgconfig file if the prefix is `/usr/local` and the includedir is `/usr/local/include` it should be represented as `${prefix}/include`, but if the includedir is something else like say `/opt/include` which doesn't exist in the prefix it should be represented as `/opt/include`. And libdir would be the same, except it uses `${exec_prefix}` instead.

In my script it compares the value of the path with the prefix, where if it matches it uses `${prefix}` or `${exec_prefix} and otherwise takes the value as given.

In JGEMU the usage is like this.
```
override PKGCONF_SH := $(wildcard $(SOURCEDIR)/lib/pkgconf.sh)

ifneq ($(PKGCONF_SH),)
        override PKGCONFLIBDIR := \
                $(shell $(PKGCONF_SH) "$(EXEC_PREFIX)" "$(LIBDIR)" exec_)
        override PKGCONFINCDIR := \
                $(shell $(PKGCONF_SH) "$(PREFIX)" "$(INCLUDEDIR)")
endif

ifeq ($(PREFIX), $(EXEC_PREFIX))
        override PKGCONFEXECDIR := $${prefix}
else
        override PKGCONFEXECDIR := $(EXEC_PREFIX)
endif
```
Also `exec_prefix` should default to the value of `prefix`, so `EXEC_PREFIX ?= $(PREFIX)`. Where `$(PREFIX)/bin` would be the `BINDIR`. Take a look at the autotools `./configure --help` for example.
```
Installation directories:
  --prefix=PREFIX         install architecture-independent files in PREFIX
                          [/usr/local]
  --exec-prefix=EPREFIX   install architecture-dependent files in EPREFIX
                          [PREFIX]

By default, `make install' will install all the files in
`/usr/local/bin', `/usr/local/lib' etc.  You can specify
an installation prefix other than `/usr/local' using `--prefix',
for instance `--prefix=$HOME'.

For better control, use the options below.

Fine tuning of the installation directories:
  --bindir=DIR            user executables [EPREFIX/bin]
  --sbindir=DIR           system admin executables [EPREFIX/sbin]
  --libexecdir=DIR        program executables [EPREFIX/libexec]
  --sysconfdir=DIR        read-only single-machine data [PREFIX/etc]
  --sharedstatedir=DIR    modifiable architecture-independent data [PREFIX/com]
  --localstatedir=DIR     modifiable single-machine data [PREFIX/var]
  --runstatedir=DIR       modifiable per-process data [LOCALSTATEDIR/run]
  --libdir=DIR            object code libraries [EPREFIX/lib]
  --includedir=DIR        C header files [PREFIX/include]
  --oldincludedir=DIR     C header files for non-gcc [/usr/include]
  --datarootdir=DIR       read-only arch.-independent data root [PREFIX/share]
  --datadir=DIR           read-only architecture-independent data [DATAROOTDIR]
  --infodir=DIR           info documentation [DATAROOTDIR/info]
  --localedir=DIR         locale-dependent data [DATAROOTDIR/locale]
  --mandir=DIR            man documentation [DATAROOTDIR/man]
  --docdir=DIR            documentation root [DATAROOTDIR/doc/pkgconf]
  --htmldir=DIR           html documentation [DOCDIR]
  --dvidir=DIR            dvi documentation [DOCDIR]
  --pdfdir=DIR            pdf documentation [DOCDIR]
  --psdir=DIR             ps documentation [DOCDIR]
```
Where values like `bindir` and `libdir` use the `exec_prefix` while `includedir` and `datarootdir` use `prefix`.

Reply:
OK; I didn't know we could override the pkg-config variables such as `prefix`, but that's a possibility according to pkg-config's man page:
```
       --define-variable=VARIABLENAME=VARIABLEVALUE
              This sets a global value for a variable, overriding the value in
              any .pc files. Most packages define the variable  "prefix",  for
              example, so you can say:
                $ pkg-config --print-errors --define-variable=prefix=/foo \
                             --variable=prefix glib-2.0
                /foo
``` 

So I've implemented something that should do the same thing as your shell script but in Make itself instead. See the latest revision, which includes the diff below:

```
modified   Makefile
@@ -42,17 +42,49 @@ ifeq ($(PLATFORM),windows32)
 NULL := NUL
 endif
 
+# Strip a given prefix from a string.
+# arg1: The prefix to strip.
+# arg2: The text containing the prefix.
+# Return $text with $prefix stripped, else nothing.
+define strip_prefix =
+$(let stripped,$(subst $(1),,$(2)),$\
+  $(shell test "$(1)$(stripped)" = "$(2)" && echo $(stripped)))
+endef
+
+# Simplify a path for use with pkg-config, by replacing $prefix with '${prefix}'
+# arg1: prefix, e.g. an installation prefix, such as /usr/local.
+# arg2: pkgdir, e.g. a path such as $datadir, $bindir, etc.
+# arg3: pkg-config variable name, e.g. 'prefix' or 'exec_prefix'.
+#       Defaults to 'prefix'.
+define simplify_pkgconf_dir =
+$(let stripped,$(call strip_prefix,$(1),$(2)),$\
+  $(if stripped,$${$(or $(3),prefix)}$(stripped),$(2)))
+endef
+
 ifneq ($(shell which xdg-open 2> $(NULL))$(FREEDESKTOP),)
 # Running on an FreeDesktop environment, configure for (optional) installation
 DESTDIR ?= 
 PREFIX ?= /usr/local
-EXEC_PREFIX ?= $(PREFIX)/bin
-INCLUDEDIR ?= $(PREFIX)/include
-LIBDIR ?= $(PREFIX)/lib
 DATA_DIR ?= $(PREFIX)/share/sameboy/
 FREEDESKTOP ?= true
 endif
 
+# Autoconf-style conventionally named variables.
+prefix ?= $(PREFIX)
+exec_prefix ?= $(prefix)
+includedir ?= $(prefix)/include
+bindir ?= $(exec_prefix)/bin
+libdir ?= $(exec_prefix)/lib
+datadir ?= $(DATA_DIR)
+
+# Prettified variants for use in the pkg-config file.
+override PKGCONF_EXEC_PREFIX = \
+	$(call simplify_pkgconf_dir,$(prefix),$(exec_prefix))
+override PKGCONF_INCLUDEDIR = \
+	$(call simplify_pkgconf_dir,$(prefix),$(includedir))
+override PKGCONF_LIBDIR = \
+	$(call simplify_pkgconf_dir,$(exec_prefix),$(libdir),exec_prefix)
+
 default: $(DEFAULT)
 
 CFLAGS += -fPIC			# required for the shared library
@@ -845,11 +877,11 @@ $(LIBDIR)/libsameboy.so: $(LIBDIR)/libsameboy.o
 $(PKGCONF_FILE): sameboy.pc.in
 	-@$(MKDIR) -p $(dir $@)
 	-@rm -f $@
-	sed -e 's,@PREFIX@,$(PREFIX),' \
-	 -e 's/@VERSION@/$(VERSION)/' \
-	 -e 's,@EXEC_PREFIX@,$(EXEC_PREFIX),' \
-	 -e 's,@INCLUDEDIR@,$(INCLUDEDIR),' \
-	 -e 's,@LIBDIR@,$(LIBDIR),' $< > $@
+	sed -e 's,@prefix@,$(prefix),' \
+	 -e 's/@version@/$(VERSION)/' \
+	 -e 's,@exec_prefix@,$(PKGCONF_EXEC_PREFIX),' \
+	 -e 's,@includedir@,$(PKGCONF_INCLUDEDIR),' \
+	 -e 's,@libdir@,$(PKGCONF_LIBDIR),' $< > $@
 
 pkgconf: $(PKGCONF_FILE)
 
modified   sameboy.pc.in
@@ -1,10 +1,10 @@
-prefix=@PREFIX@
-exec_prefix=@EXEC_PREFIX@
-includedir=@INCLUDEDIR@
-libdir=@LIBDIR@
+prefix=@prefix@
+exec_prefix=@exec_prefix@
+includedir=@includedir@
+libdir=@libdir@
 
 Name: sameboy
 Description: The SameBoy library
-Version: @VERSION@
+Version: @version@
 Cflags: -I${includedir}/sameboy
 Libs: -L${libdir} -lsameboy
```

This produces:
```
$ cat build/lib/pkgconfig/sameboy.pc 
prefix=/usr/local
exec_prefix=${prefix}
includedir=${prefix}/include
libdir=${exec_prefix}/lib

Name: sameboy
Description: The SameBoy library
Version: 0.16.6
Cflags: -I${includedir}/sameboy
Libs: -L${libdir} -lsameboy
```